llama-rs/src/llama.rs

```diff
@@ -591,54 +630,59 @@ impl LlamaModel {
     }
 
     pub fn inference_with_prompt<'a>(
-        &self,
+        &mut self,
         vocab: &'a GptVocab,
         params: &InferenceParams,
         prompt: &str,
         rng: &mut impl rand::Rng,
+        stop_after_prompt: bool,
         callback: impl Fn(OutputToken<'a>),
     ) {
-        let embd_inp = self.tokenize(vocab, prompt, true);
-        let mut logits = Vec::new();
+        let beginning_of_sentence = self.n_past == 0;
+        let embd_inp = self.tokenize(vocab, prompt, beginning_of_sentence);
```

Comment:
One sneaky issue I found was that we don't want to feed the beginning of sentence token when restoring from memory, because that causes the system to insert that token into the memory and then the system ignores any tokens before it.